Add java compiler error reporting release note (#30242)
gradle/gradle: platforms/documentation/docs/src/docs/release/notes.md

For details, see the full [compatibility documentation](userguide/compatibility.html#java).

<a name="java-compiler-error-rendering"></a>
### Java compiler errors in the failure report

> From now on, Gradle also includes Java compilation failures in the report at the end of the build.

Reviewer: I think we should start with the context. What's the problem we wanted to solve:
- in previous Gradle versions, the build solely relied on the compiler output
- If build failed, the message just had an advice to find it in the build log
- For large builds and for CI this means tedious scrolling finding the corresponding compile task output

Author: FYI, I've wanted to tell this story, but I was unsure how much storytelling we want to do in the release notes. I guess by your comment, you also think it would be worthwhile, so I will expand upon the details

Reviewer: I've added @lkasso to this review explicitly to help us with the content.

Author: Pushed the new text, then resolve or comment

Author: Left CI story out, my bad

Author: Though new text might be enough, and we don't need to shoehorn in CI there.

This will accompany the already existing errors reported when the task is executed.
This improvement helps to identify compilation problems faster; with long executions, it can be hard to find the root cause of a compilation failure.

The following is a sample output caused by a simple compilation failure:
```
* What went wrong:
Execution failed for task ':project1:compileJava'.
> Compilation failed; see the compiler output below.

Java compilation error (compilation:java:java-compilation-error)
  sample-project/src/main/java/Problem.java:6: error: incompatible types: int cannot be converted to String
          String a = 1;
                     ^
```

If any warning happens during the compilation, it will also be included in the report:
```
* What went wrong:
Execution failed for task ':project1:compileJava'.
> Compilation failed; see the compiler output below.

> Java compilation warning (compilation:java:java-compilation-warning)

Reviewer: I'm not sure if we discussed this deeply; but I don't think we should have this line in the output. The ID (`compilation:java:java-compilation-warning`) is an internal detail for the problems API and the display name (`Java compilation warning`) contains information already present in the next line.

    sample-project/src/main/java/Problem1.java:6: warning: [cast] redundant cast to String
                                                  ^

Author: I understand your concern, but I don't find the discussion about this on the PR, and you didn't block the PR neither.

I can still fix this, just noting that this misunderstanding could've been avoided.

  sample-project/src/main/java/Problem1.java:6: warning: [cast] redundant cast to String
          var warning = (String)"warning";
                        ^
Java compilation error (compilation:java:java-compilation-error)
  sample-project/src/main/java/Problem2.java:6: error: incompatible types: int cannot be converted to String
          String a = 1;
                     ^
```

> A limitation of this feature is that warnings are only reported when a failure occurs.

Reviewer: It's not really a limitation, it's the scope of the feature. I propose we soften the sentence a little bit, by saying
```suggestion
Note, that the output remains unchanged when the Java compilation task finishes with warnings only. 
```
Also, this should be the last sentence of the release notes.

Author: I think the solution is somewhere in the middle: we should denote that if there is no failure, there are no warnings displayed.

Author: Pushed the new text, then resolve or comment

> We are working on additional rendering methods, which would display warnings even when the compilation is successful.

Reviewer: This belongs to the roadmap.

```suggestion
```

Author: How can people find the roadmap?

Reviewer: https://blog.gradle.org/roadmap-announcement

Author: Also, I want the users to know that we still care about reporting warnings when only warnings happen, but it is not ready yet. If you think from the user perspective, this story is not yet complete, and I want to emphasize that further features are coming.

Author: > https://blog.gradle.org/roadmap-announcement

Maybe I haven't asked the correct question: how the users should know that this is part of a roadmap if we don't include the roadmap.
Do yo want me to include it?

Author: Am I wrong not to see a roadmap item for Problems API reporting on the roadmap? I can see the Java one, but I don't see the universal reporting features.

Reviewer: So my comment was about the fact that we should talk about stuff that we have in the release notes and not promise anything. Unlike the published release notes, we can adjust the roadmap later.

Author: Might not have enough time to make PR

Reviewer: Maybe we should only mention that we'll have more enhancements as part of the [`Clean and actionable error reporting`](https://github.com/gradle/build-tool-roadmap/issues/49) roadmap item. 

Also, I'm happy to adjust the description on the roadmap based on your feedback.

Author: That being said, the solution uses a common rendering class: https://github.com/gradle/gradle/blob/282f62db0ac5e4f59c442229bafcb7feb0ff4593/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/ProblemRenderer.java#L32

I didn't make any strategy pattern here on differentiating the rendering, as I believed we agreed on an uniform look between a (coming) problem console reporting, and this report

Author: Pushed the new test, PTAL

> 

Reviewer: I'm not sure on this but we might want to reiterate that this feature is built on top of the Problems API. The important bit for me is that we expose structural information consistently to many consumers: build scans, IDEs, and the console.  We could also add a link to the 8.9 release notes where we first mentioned the compiler integration.

Author: I don't think this is relevant here for the users reading the release notes, especially as we already had the integration enabled in 8.9.

This is our internal adoption of using the data in a new format.

Reviewer: Ok.

For details, see the full [compatibility documentation](userguide/compatibility.html#java).

<a name="java-compiler-error-rendering"></a>
### Java compiler errors in the failure report

> In previous Gradle versions, finding the reason of a failed compilation was a suboptimal experience: the only pointer we could give to the user was to scroll back, and look for the failed task's output, containing the compiler failure.

Reviewer 2: ```suggestion
In previous Gradle versions, identifying the cause of a failed compilation was often cumbersome. 
Users had to scroll through lengthy build logs to find the task output containing the compiler error.
```

> This user experience wasn't great; build logs can be extremely long, and our support for multiple failures (when using the `--continue` flag) could complicate the situation further, making the identification of the exact failure reason challenging.

Reviewer 2: ```suggestion
This process was particularly challenging when using the `--continue` flag, as handling multiple failures could make it difficult to pinpoint the exact cause of an error.
```

Author: I think we lost some meaning there:
 - The key point is that this feature helps with long logs
 - I've tried to convey that if for example you have long logs, and use `--continue` (tbh, not a super commonly used feature), you might have to spend a decent time on finding the root issue of your compilation

> Gradle now gained the ability to collect, and report per-task failures in the bottom "What went wrong" segment.

Reviewer 2: ```suggestion
Gradle now has the ability to collect and report per-task failures in the "What went wrong" section at the bottom of the build log.
```

> This report supplements the usual task output, and aims to give a new way to identify problems quicker and easier.

Reviewer 2: ```suggestion
This enhancement supplements the usual task output and provides a more straightforward way to identify issues.
```

> For example, a simple failure in the bottom of the log will show up as:

Reviewer 2: ```suggestion
For example, a simple failure now appears in the bottom of the log as:
```

```
* What went wrong:
Execution failed for task ':project1:compileJava'.
> Compilation failed; see the compiler output below.

Java compilation error (compilation:java:java-compilation-error)
  sample-project/src/main/java/Problem.java:6: error: incompatible types: int cannot be converted to String
          String a = 1;
                     ^
```

> If any warning happens during the compilation, it will also be included in the report:

Reviewer 2: ```suggestion
Warnings during compilation are also included in the report:

```

```
* What went wrong:
Execution failed for task ':project1:compileJava'.
> Compilation failed; see the compiler output below.

Java compilation warning (compilation:java:java-compilation-warning)
  sample-project/src/main/java/Problem1.java:6: warning: [cast] redundant cast to String
          var warning = (String)"warning";
                        ^
Java compilation error (compilation:java:java-compilation-error)
  sample-project/src/main/java/Problem2.java:6: error: incompatible types: int cannot be converted to String
          String a = 1;
                     ^
```

> Note that the current solution reports upon _failures_. If only warnings happen (and no `-Werror` is set), this report will not be visible.

Reviewer 2: ```suggestion
Note that this report only appears when there are _failures_. If only warnings occur (and `-Werror` is not set), this report will not be shown.
```

Author: Opps, good catch with the _not_! 🎉

> The feature also works with the [`--continue`](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continue_build_on_failure) flag, and the bottom report will contain a per-task report of all the compilation failures.

Reviewer 2: ```suggestion
This feature also works with the [`--continue`](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continue_build_on_failure) flag, providing a per-task report of all compilation failures at the bottom of the log.
```

```
* What went wrong:
Execution failed for task ':project1:compileJava'.
> Compilation failed; see the compiler output below.

Java compilation error (compilation:java:java-compilation-error)
  sample-project/src/main/java/Problem.java:6: error: incompatible types: int cannot be converted to String
          String a = 1;
                     ^
```

If any warning happens during the compilation, it will also be included in the report:
```
* What went wrong:
Execution failed for task ':project1:compileJava'.
> Compilation failed; see the compiler output below.

Java compilation warning (compilation:java:java-compilation-warning)
  sample-project/src/main/java/Problem1.java:6: warning: [cast] redundant cast to String
          var warning = (String)"warning";
                        ^
Java compilation error (compilation:java:java-compilation-error)
  sample-project/src/main/java/Problem2.java:6: error: incompatible types: int cannot be converted to String
          String a = 1;
                     ^
```

Note that the current solution reports upon _failures_. If only warnings happen (and no `-Werror` is set), this report will not be visible.

> The feature also works with the [`--continue`](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continue_build_on_failure) flag, and the bottom report will contain a per-task report of all the compilation failures.

Reviewer 3: This should not use an absolute URL.  It needs to be relative like:
```suggestion
The feature also works with the [`--continue`](userguide/command_line_interface.html#sec:continue_build_on_failure) flag, and the bottom report will contain a per-task report of all the compilation failures.
```
